library/core/src/num/int_macros.rs

```diff
@@ -1580,7 +1580,18 @@ macro_rules! int_impl {
             if self < 0 {
                 None
             } else {
-                Some((self as $UnsignedT).isqrt() as Self)
+                let result = crate::num::int_sqrt::$ActualT(self as $ActualT) as $SelfT;
+
+                // SAFETY: Inform the optimizer that square roots of
+                // nonnegative integers are nonnegative and what the maximum
+                // result is.
```

Comment:
Maybe

```suggestion
                // Inform the optimizer that the square root of nonnegative integers is bounded.
                //
                // SAFETY: `self` is checked above to be nonnegative, hence `sqrt(self) >= 0`. `sqrt(x)` is
                // monotonically increasing ∀ x >= 0, so ∵ `x <= x_max`, `sqrt(x) <= sqrt(x_max)`
```

Feel free to make that less messy, just separate the action (informing the optimizer) from the precondition justification (why the condition to `assert_unchecked` must be true)

Reply:
Is this too verbose for the `NonZero` variant?

```rust
            // SAFETY: Inform the optimizer what the range of outputs is.
            //
            // Integer square root is a monotonically nondecreasing
            // function, which means that increasing the input will never
            // cause the output to decrease.
            //
            // The minimum input is 1. The maximum input is `<$Int>::MAX`.
            //
            // When n is 1, sqrt(n) is 1. If n increases above 1, sqrt(n)
            // can't decrease below 1, so sqrt(n) can't decrease below 1 no
            // matter what n is.
            //
            // When n is below `<$Int>::MAX`, sqrt(n) can't decrease at all
            // when you increase n to `<$Int>::MAX`, so sqrt(n) can't be above
            // sqrt(`<$Int>::MAX`) no matter what n is.
```

Reply:
Lol indeed verbose, but also accurate so I don't think there is a problem.